baselines/fedstar/fedstar/models.py

```diff
@@ -0,0 +1,426 @@
+        super().__init__()
+        self.model = Network(
+            num_classes=num_classes, dropout_rate=dropout_rate
+        ).get_network()
+        self.confidence = 0.5
+        self.aux_loss_weight = aux_loss_weight
+        self.history = History()
+        self.optimizer = None
+        self.loss_fn = None
+
+    # pylint: disable=arguments-differ
+    def compile(self, optimizer):
+        """Configure the model for training.
+
+        Sets the optimizer and loss function for the model.
+
+        Parameters
+        ----------
+        - optimizer: The optimizer to use for training the model.
+        """
+        super().compile()
+        self.optimizer = optimizer
+        self.loss_fn = tf.nn.sparse_softmax_cross_entropy_with_logits
+
+    @staticmethod
+    def cosine_confidence_scheduler(
+        step, total_steps, confidence_min=0.5, confidence_max=0.9
+    ):
+        """Determine the confidence level for pseudo-labeling using a cosine schedule.
+
+        Gradually adjusts the confidence threshold over the course of training based
+        on the cosine function.
+
+        Parameters
+        ----------
+        - step (int): The current training step or round.
+        - total_steps (int): The total number of steps or rounds
+                             in the training process.
+        - confidence_min (float, optional): Minimum confidence threshold.
+        - confidence_max (float, optional): Maximum confidence threshold.
+
+        Returns
+        -------
+        - float: The computed confidence threshold for the current step.
+        """
+        return confidence_max - (confidence_max - confidence_min) * 0.5 * (
+            1 + np.cos(step / total_steps * np.pi)
+        )
+
+    # pylint: disable=too-many-arguments, too-many-locals, arguments-differ
+    def fit(
+        self, train_datasets, epochs, c_round, rounds, num_batches=None, T=1, verbose=0
+    ):
+        """Trains the model on given datasets for a specified number of epochs.
+
+        This custom training loop handles both labeled and unlabeled data, applying
+        pseudo-labeling for semi-supervised learning.
+
+        Parameters
+        ----------
+        - train_datasets (tuple): A tuple containing labeled
+                                  and unlabeled training datasets.
+        - epochs (int): Number of epochs to train the model.
+        - c_round (int): Current round of training in a federated learning scenario.
+        - rounds (int): Total number of rounds in the federated learning scenario.
+        - num_batches (int, optional): Number of batches per epoch.
+        - T (float, optional): Temperature parameter for softmax in pseudo-labeling.
+        - verbose (int, optional): Verbosity mode.
+
+        Returns
+        -------
+        - History: An object containing training history information.
+        """
+        l_dataset, u_dataset = train_datasets
+        l_num_batch = len(l_dataset)

```

Comment:
this fails when using `table4` experiments with `speech_commands`. I suspect it's due to the balancing. It should be possible to achieve the same functionality since you already know what the `l_per` and `u_per` values are and what's the total dataset length of the client right?